tests/library/test_json.py

```diff
@@ -363,6 +363,140 @@ def test_bad_object(self, bad_string, failure_byte):
+            (5, 5, [42, True, "hello", "world", "test"]),  # Exactly meets minItems and maxItems.
+            (0, 10, [42, True] + ["extra"] * 8),  # Exactly meet large number of extra items
+        ]
+    )
+    def test_good_with_prefix_and_items(self, min_items, max_items, target_obj):
+        schema_obj = {
+            "prefixItems": self.prefix_schema_obj,
+            "items": self.items_schema_obj,
+            "minItems": min_items,
+            "maxItems": max_items,
+            "type": "array",
+        }
+        _generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "min_items, max_items, target_obj",
+        [
+            (0, 0, []), # None allowed, none provided
+            (0, 2, []), # Some allowed, none provided
+            (1, 2, [42, True]), # All prefix items, no extra allowed
+            (2, 2, [42, True]), # Exactly match min, max
+            (1, 3, [42]),  # Single prefix item, extra allowed
+            (1, 3, [42, True]), # All prefix items, extra allowed
+        ]
+    )
+    def test_good_with_prefix(self, min_items, max_items, target_obj):
+        schema_obj = {
+            "prefixItems": self.prefix_schema_obj,
+            "minItems": min_items,
+            "maxItems": max_items,
+            "type": "array",
+        }
+        _generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "min_items, max_items, target_obj",
+        [
+            (0, 0, []), # None allowed, none provided
+            (0, 2, []), # Some allowed, none provided
+            (1, 2, ["hello"]),  # Single item, more allowed
+            (1, 2, ["hello", "world"]),  # Meet max
+            (3, 3, ["hello", "world", "extra"]), # Exactly match min, max
+            (0, 8, ["extra"]*8),  # Large number of items
+        ]
+    )
+    def test_good_with_items(self, min_items, max_items, target_obj):
+        schema_obj = {
+            "items": self.items_schema_obj,
+            "minItems": min_items,
+            "maxItems": max_items,
+            "type": "array",
+        }
+        _generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "min_items, max_items, bad_obj, failure_byte",
+        [
+            (1, 4, [42, "string_not_bool", "hello", "extra"], b'"'),  # Second item does not match prefix schema
+            (0, 3, [42, True, 100], b"1"),  # Last item does not match general item schema
+            (3, 5, [42, True, "valid", "extra1", "extra2", "too_many"], b","),  # Exceeds maxItems
+            (2, 3, [42], b"]"),  # Not enough items
+            (1, 1, [42, True], b","),  # Too many items for maxItems
+            (0, 0, [42, True, "str"], b"4"),  # maxItems set to 0, but array is not empty
+            (3, 5, [42, True], b"]"),  # Array has one fewer item than required by minItems
+        ]
+    )
+    def test_bad_with_prefix_and_items(self, min_items, max_items, bad_obj, failure_byte):
```

Comment:
A long shot (and perhaps not worth worrying about right now), but is there any way of getting a more useful error message than just the failure byte?

Reply:
Adding byte position would be great, maybe alongside the bytes that have been consumed (for redundancy). Also adding "valid next bytes" that would have been accepted instead of the failure byte would be nice.

I'd be happy to take a look at the parser code to see if adding these wouldn't be too difficult

Reply:
Definitely a separate PR :)

Reply:
Added an issue to track this: https://github.com/guidance-ai/guidance/issues/739